feat(expression-input): implement basic calculator bar

orbitalsqwib opened the pull request:
## Problem

Users are unsure how to type in math symbols or how to reference questions, which is affecting reach and retention. To solve this, a calculator bar can help users visualize and see what math symbols they can use, as well as identify the reference symbol

Closes #472

## Solution

Implement a basic calculator bar for expression inputs, displayed on input focus.

**Features**:

- implement calculator bar into expression inputs and enable them to add/replace content within the input.

## Screenshots

![Screenshot 2021-05-25 at 9 50 22 PM](https://user-images.githubusercontent.com/21305518/119509425-387d8080-bda3-11eb-94eb-ad40a154afd9.png)
![Screenshot 2021-05-25 at 9 50 41 PM](https://user-images.githubusercontent.com/21305518/119509488-429f7f00-bda3-11eb-924e-0c276f9367c3.png)

## Tests

- [ ] Check that calculator bar can add operators into the expression input
- [ ] Check that entering the `@` operand through the calculator bar brings up the query replacement dropdown
- [ ] Check that selecting a selection of characters in the expression input, then subsequently clicking on a operand in the calculator bar will replace that selection with the new character (select and replace)

Commit:
refactor(expression-input): make downshift controlled

- remove control of input-value and is-open from downshift and use a
  single state as a source of truth for the entire component
- minimise the use of get-input-props to prepare for textarea
- implement useEffects to ensure data is kept up to date
- implement controlled selection using layout effects
- always render unordered-list to ensure that menu props is always
  called, thereby retaining accessibility properties
- clean up code

Commit:
feat(calculator-bar): first working version, input-scoped

Commit:
feat(calculator-bar): use flex wrap instead of scroll on overflow

Commit:
Merge branch 'develop' into feat/input-scoped-calculator-bar

prestonlimlianjie commented on src/client/components/builder/logic/CalculatorBar.tsx:
```diff
@@ -0,0 +1,36 @@
+  onClick,
+  ...props
+}) => {
+  const operands = '@+-×÷()^%>≥<≤=≠'.split('')
```
We should create a map of operator to mappedOperator so that we won't run the risk of messing up the indices when we add/remove more operators in the future, e.g.
```
const operatorMap = { 
  "@": "@",
  "=": "==",
   ...
} 
```

P.S. it should be **operator** not **operand** - **operand** refers to the object of the operation, i.e. in `3 + 6`, `3` and `6` are the **operands** while `+` is the **operator**

prestonlimlianjie commented on src/client/components/builder/logic/ExpressionInput.tsx:
```diff
@@ -186,93 +226,127 @@ export const ExpressionInput: FC<ExpressionInputProps> = ({
     return inputStr || ''
   }
 
+  const handleCalculatorChange = (operand: string) => {
```
Should be **operator** not **operand**

prestonlimlianjie commented on src/client/components/builder/logic/ExpressionInput.tsx:
```diff
@@ -51,9 +52,48 @@ export const ExpressionInput: FC<ExpressionInputProps> = ({
   const inputRef = useRef<HTMLInputElement>(null)
+  const wrapperRef = useRef<HTMLDivElement>(null)
+
+  // syncs input value with subsequent value updates
```
The code in lines 61-80 is duplicated from line 98 onwards

orbitalsqwib replied:
Thanks for catching this! I must've been tired when I was resolving merge conflicts with develop...

prestonlimlianjie reviewed (changes requested):
The PR looks good - some minor changes requested below

Commit:
refactor(calculator-bar): map operators using map instead of array

- fixed the term 'operand' -> 'operator'
- fixed a duplicated piece of code in expression input from a messy git
  merge conflict resolution

orbitalsqwib commented:
I've implemented the map as a Map instead of an Object so we can use iterators (and keep the order consistent without having to repeat ourselves with an array).

Let me know if this is okay with you!

prestonlimlianjie reviewed (approved):
lgtm!
